File: packages/shared/src/network-proxy.ts
/**
 * User-configurable outbound proxy (D340 / ADR 0177).
 *
 * Pure parse/validate helpers shared by the renderer, Electron main, and the
 * agent sidecar. Applying a proxy to Chromium sessions, process env, or
 * undici lives in the process that owns those surfaces.
 */

export const NETWORK_PROXY_MODES = ["system", "direct", "custom"] as const;
export type NetworkProxyMode = (typeof NETWORK_PROXY_MODES)[number];

export const NETWORK_PROXY_SCHEMES = [
  "http",
  "https",
  "socks",
  "socks5",
  "socks5h",
] as const;
export type NetworkProxyScheme = (typeof NETWORK_PROXY_SCHEMES)[number];

/**
 * Default bypass list for custom mode.
 *
 * Loopback plus the private ranges a user's own LAN devices live in. A proxy is
 * for reaching the public internet; sending `192.168.0.0/16` through it is how
 * a local model server, NAS or MCP endpoint stops answering. `<local>` keeps
 * Chromium's own single-label rule.
 */
export const DEFAULT_NETWORK_PROXY_BYPASS =
  "localhost,127.0.0.1,::1,<local>,10.0.0.0/8,172.16.0.0/12,192.168.0.0/16,169.254.0.0/16";

export const PROXY_ENV_KEYS = [
  "HTTP_PROXY",
  "HTTPS_PROXY",
  "ALL_PROXY",
  "NO_PROXY",
  "http_proxy",
  "https_proxy",
  "all_proxy",
  "no_proxy",
  "NODE_USE_ENV_PROXY",
] as const;

export type NetworkProxySettings = {
  mode: NetworkProxyMode;
  /** Required when `mode` is `custom`. Canonical proxy URL. */
  url?: string;
  /**
   * Comma-separated bypass list. Absent uses
   * {@link DEFAULT_NETWORK_PROXY_BYPASS} for custom mode.
   */
  bypass?: string;
};

export type ParsedProxyUrl = {
  href: string;
  scheme: NetworkProxyScheme;
  host: string;
  port: number | null;
  username: string;
  password: string;
  isSocks: boolean;
};

const PROXY_URL_MAX_LENGTH = 2048;

const SCHEME_SET = new Set<string>(NETWORK_PROXY_SCHEMES);

export function isNetworkProxyMode(value: unknown): value is NetworkProxyMode {
  return (
    typeof value === "string" &&
    (NETWORK_PROXY_MODES as readonly string[]).includes(value)
  );
}

export function normalizeNetworkProxy(value: unknown): NetworkProxySettings {
  const record =
    value && typeof value === "object" && !Array.isArray(value)
      ? (value as Record<string, unknown>)
      : null;
  const mode = isNetworkProxyMode(record?.mode) ? record.mode : "system";
  const parsed =
    typeof record?.url === "string" ? parseProxyUrl(record.url) : null;
  const bypass =
    typeof record?.bypass === "string" ? record.bypass.trim() : "";
  const settings: NetworkProxySettings = { mode };
  if (parsed?.ok) settings.url = parsed.value.href;
  else if (typeof record?.url === "string" && record.url.trim()) {
    settings.url = record.url.trim();
  }
  if (bypass) settings.bypass = bypass;
  // A build before the network policy owned it stored `allowFakeIp` here. The
  // key is ignored, not rejected: the proxy decides where traffic goes, and
  // `networkPolicy.mode` decides what the app tolerates from the route.
  return settings;
}

export function effectiveProxyBypass(settings: NetworkProxySettings): string {
  const bypass = settings.bypass?.trim();
  return bypass || DEFAULT_NETWORK_PROXY_BYPASS;
}

export type ParseProxyUrlResult =
  | { ok: true; value: ParsedProxyUrl }
  | { ok: false; error: string };

export function parseProxyUrl(raw: string): ParseProxyUrlResult {
  const trimmed = raw.trim();
  if (!trimmed) {
    return { ok: false, error: "proxy URL is required" };
  }
  if (trimmed.length > PROXY_URL_MAX_LENGTH) {
    return { ok: false, error: "proxy URL is too long" };
  }
  if (/\s/.test(trimmed)) {
    return { ok: false, error: "proxy URL must not contain whitespace" };
  }
  let url: URL;
  try {
    url = new URL(trimmed);
  } catch {
    return { ok: false, error: "proxy URL is invalid" };
  }
  const scheme = url.protocol.replace(/:$/, "").toLowerCase();
  if (!SCHEME_SET.has(scheme)) {
    return {
      ok: false,
      error: "proxy scheme must be http, https, socks, socks5, or socks5h",
    };
  }
  const host = url.hostname.trim();
  if (!host) {
    return { ok: false, error: "proxy URL must include a host" };
  }
  let port: number | null = null;
  if (url.port) {
    const parsedPort = Number(url.port);
    if (
      !Number.isInteger(parsedPort) ||
      parsedPort < 1 ||
      parsedPort > 65535
    ) {
      return { ok: false, error: "proxy port must be between 1 and 65535" };
    }
    port = parsedPort;
  }
  let username: string;
  let password: string;
  try {
    username = decodeURIComponent(url.username);
    password = decodeURIComponent(url.password);
  } catch {
    return { ok: false, error: "proxy credentials are invalid" };
  }
  const isSocks = scheme.startsWith("socks");
  const href = formatProxyHref({
    scheme: scheme as NetworkProxyScheme,
    host,
    port,
    username,
    password,
  });
  return {
    ok: true,
    value: {
      href,
      scheme: scheme as NetworkProxyScheme,
      host,
      port,
      username,
      password,
      isSocks,
    },
  };
}

function formatProxyHref(parts: {
  scheme: NetworkProxyScheme;
  host: string;
  port: number | null;
  username: string;
  password: string;
}): string {
  const host =
    parts.host.includes(":") && !parts.host.startsWith("[")
      ? `[${parts.host}]`
      : parts.host;
  const port = parts.port ? `:${parts.port}` : "";
  let auth = "";
  if (parts.username || parts.password) {
    auth = `${encodeURIComponent(parts.username)}`;
    if (parts.password) auth += `:${encodeURIComponent(parts.password)}`;
    auth += "@";
  }
  return `${parts.scheme}://${auth}${host}${port}`;
}

export function validateNetworkProxy(
  value: unknown,
): { ok: true; value: NetworkProxySettings } | { ok: false; error: string } {
  const settings = normalizeNetworkProxy(value);
  if (settings.mode !== "custom") {
    const next: NetworkProxySettings = { mode: settings.mode };
    return { ok: true, value: next };
  }
  const parsed = parseProxyUrl(settings.url ?? "");
  if (!parsed.ok) return parsed;
  const next: NetworkProxySettings = {
    mode: "custom",
    url: parsed.value.href,
  };
  if (settings.bypass?.trim()) next.bypass = settings.bypass.trim();
  return { ok: true, value: next };
}

export type ChromiumProxyConfig =
  | { mode: "system" }
  | { mode: "direct" }
  | { proxyRules: string; proxyBypassRules: string };

export function chromiumProxyConfig(
  settings: NetworkProxySettings,
): ChromiumProxyConfig {
  if (settings.mode === "direct") return { mode: "direct" };
  if (settings.mode !== "custom") return { mode: "system" };
  const parsed = parseProxyUrl(settings.url ?? "");
  if (!parsed.ok) return { mode: "direct" };
  return {
    proxyRules: chromiumProxyRules(parsed.value),
    proxyBypassRules: effectiveProxyBypass(settings),
  };
}

/**
 * Chromium proxy rule for one parsed proxy URL.
 *
 * Chromium understands `http`, `https`, `socks` (SOCKS5), `socks4` and
 * `socks5` proxy rules, but not curl's `socks5h` spelling: `session.setProxy`
 * accepts the unsupported rule set without an error, then resolves to no
 * proxy at all and every request fails with `net::ERR_NO_SUPPORTED_PROXIES`.
 * Chromium's SOCKS5 already hands the hostname to the proxy, so `socks5h`
 * maps onto `socks5` without changing where names are resolved.
 *
 * Userinfo is also stripped. Chromium's proxy-rule parser rejects
 * `scheme://user:pass@host:port` with the same `ERR_NO_SUPPORTED_PROXIES`
 * failure (issue #490). Electron main points Chromium at a loopback SOCKS5
 * relay that injects those credentials; Node and curl keep the canonical
 * href from {@link parseProxyUrl}.
 */
export function chromiumProxyRules(proxy: ParsedProxyUrl): string {
  return formatProxyHref({
    scheme: proxy.scheme === "socks5h" ? "socks5" : proxy.scheme,
    host: proxy.host,
    port: proxy.port,
    username: "",
    password: "",
  });
}

/** True when Chromium cannot carry the credentials in `proxyRules`. */
export function proxyHasCredentials(proxy: ParsedProxyUrl): boolean {
  return Boolean(proxy.username || proxy.password);
}

/** Bypass list suitable for `NO_PROXY` / curl `--noproxy` (no Chromium `<local>`). */
export function envProxyBypass(settings: NetworkProxySettings): string {
  return effectiveProxyBypass(settings)
    .split(",")
    .map((part) => part.trim())
    .filter((part) => part && part !== "<local>")
    .join(",");
}

/**
 * Env assignments for a child process. `null` means the key should be
 * removed so a direct/custom switch cannot leak a previous value.
 */
export function proxyEnvAssignments(
  settings: NetworkProxySettings,
): Record<string, string | null> {
  const assignments: Record<string, string | null> = {};
  for (const key of PROXY_ENV_KEYS) assignments[key] = null;
  if (settings.mode !== "custom") return assignments;
  const parsed = parseProxyUrl(settings.url ?? "");
  if (!parsed.ok) return assignments;
  const href = parsed.value.href;
  const bypass = envProxyBypass(settings);
  assignments.ALL_PROXY = href;
  assignments.all_proxy = href;
  assignments.NO_PROXY = bypass;
  assignments.no_proxy = bypass;
  assignments.NODE_USE_ENV_PROXY = "1";
  if (!parsed.value.isSocks) {
    assignments.HTTP_PROXY = href;
    assignments.HTTPS_PROXY = href;
    assignments.http_proxy = href;
    assignments.https_proxy = href;
  }
  return assignments;
}

export type ProxyEnvMap = Record<string, string | undefined>;

export function snapshotProxyEnv(
  env: ProxyEnvMap,
): Record<string, string | undefined> {
  const snapshot: Record<string, string | undefined> = {};
  for (const key of PROXY_ENV_KEYS) snapshot[key] = env[key];
  return snapshot;
}

export function applyProxyEnvAssignments(
  assignments: Record<string, string | null>,
  env: ProxyEnvMap,
): void {
  for (const [key, value] of Object.entries(assignments)) {
    if (value === null) delete env[key];
    else env[key] = value;
  }
}

export function restoreProxyEnv(
  snapshot: Record<string, string | undefined>,
  env: ProxyEnvMap,
): void {
  for (const key of PROXY_ENV_KEYS) {
    const value = snapshot[key];
    if (value === undefined) delete env[key];
    else env[key] = value;
  }
}

export function stripProxyEnv(env: ProxyEnvMap): ProxyEnvMap {
  const next: ProxyEnvMap = { ...env };
  for (const key of PROXY_ENV_KEYS) delete next[key];
  return next;
}

export function redactProxyUrl(raw: string): string {
  const parsed = parseProxyUrl(raw);
  if (!parsed.ok) return raw;
  if (!parsed.value.username && !parsed.value.password) return parsed.value.href;
  return formatProxyHref({
    scheme: parsed.value.scheme,
    host: parsed.value.host,
    port: parsed.value.port,
    username: parsed.value.username ? parsed.value.username : "",
    password: parsed.value.password ? "***" : "",
  });
}

File: crates/host-core/src/network_proxy.rs
use serde_json::Value;
use std::sync::RwLock;

const PROXY_URL_MAX: usize = 2048;

#[derive(Debug, Clone, PartialEq, Eq, Default)]
pub enum ProxyMode {
    #[default]
    System,
    Direct,
    Custom {
        url: String,
        bypass: String,
    },
}

static MARKET_PROXY: RwLock<ProxyMode> = RwLock::new(ProxyMode::System);

/// Mirrors `DEFAULT_NETWORK_PROXY_BYPASS` in `packages/shared`: loopback plus
/// the private ranges a user's own LAN devices live in, so a custom proxy never
/// swallows a local model server, NAS or MCP endpoint.
const DEFAULT_BYPASS: &str =
    "localhost,127.0.0.1,::1,<local>,10.0.0.0/8,172.16.0.0/12,192.168.0.0/16,169.254.0.0/16";

pub fn apply_from_settings(value: Option<&Value>) {
    let next = proxy_from_settings(value);
    if let Ok(mut slot) = MARKET_PROXY.write() {
        *slot = next;
    }
}

pub fn proxy_from_settings(value: Option<&Value>) -> ProxyMode {
    let Some(object) = value.and_then(Value::as_object) else {
        return ProxyMode::System;
    };
    let proxy = object.get("networkProxy").and_then(Value::as_object);
    let Some(proxy) = proxy else {
        return ProxyMode::System;
    };
    match proxy
        .get("mode")
        .and_then(Value::as_str)
        .unwrap_or("system")
    {
        "direct" => ProxyMode::Direct,
        "custom" => {
            let url = proxy
                .get("url")
                .and_then(Value::as_str)
                .unwrap_or("")
                .trim()
                .to_string();
            let bypass = proxy
                .get("bypass")
                .and_then(Value::as_str)
                .map(str::trim)
                .filter(|s| !s.is_empty())
                .unwrap_or(DEFAULT_BYPASS)
                .to_string();
            if url.is_empty() {
                ProxyMode::Direct
            } else {
                ProxyMode::Custom { url, bypass }
            }
        }
        _ => ProxyMode::System,
    }
}

pub fn validate_network_proxy(value: &Value) -> Result<(), String> {
    let Some(object) = value.as_object() else {
        return Ok(());
    };
    let Some(proxy) = object.get("networkProxy") else {
        return Ok(());
    };
    if proxy.is_null() {
        return Ok(());
    }
    let Some(proxy) = proxy.as_object() else {
        return Err("networkProxy must be an object".into());
    };
    let mode = proxy
        .get("mode")
        .and_then(Value::as_str)
        .unwrap_or("system");
    match mode {
        "system" | "direct" => Ok(()),
        "custom" => {
            let url = proxy.get("url").and_then(Value::as_str).unwrap_or("");
            parse_proxy_url(url).map(|_| ())
        }
        other => Err(format!("unknown networkProxy.mode '{other}'")),
    }
}

pub fn parse_proxy_url(raw: &str) -> Result<String, String> {
    let trimmed = raw.trim();
    if trimmed.is_empty() {
        return Err("proxy URL is required".into());
    }
    if trimmed.len() > PROXY_URL_MAX {
        return Err("proxy URL is too long".into());
    }
    if trimmed.chars().any(char::is_whitespace) {
        return Err("proxy URL must not contain whitespace".into());
    }
    let Some((scheme, rest)) = trimmed.split_once("://") else {
        return Err("proxy URL must include a scheme".into());
    };
    let scheme = scheme.to_ascii_lowercase();
    if !matches!(
        scheme.as_str(),
        "http" | "https" | "socks" | "socks4" | "socks4a" | "socks5" | "socks5h"
    ) {
        return Err("proxy scheme must be http, https, or socks5".into());
    }
    let hostport = rest.split(['/', '?', '#']).next().unwrap_or(rest);
    let hostport = match hostport.rfind('@') {
        Some(at) => &hostport[at + 1..],
        None => hostport,
    };
    if hostport.is_empty() || hostport.starts_with(':') {
        return Err("proxy URL must include a host".into());
    }
    Ok(trimmed.to_string())
}

/// Extra curl arguments so marketplace downloads honor Custom/Direct without
/// mutating host-core process env (Bash must not inherit proxy credentials).
pub fn curl_proxy_args() -> Vec<String> {
    let mode = MARKET_PROXY
        .read()
        .map(|guard| guard.clone())
        .unwrap_or_default();
    match mode {
        ProxyMode::System => Vec::new(),
        ProxyMode::Direct => vec!["--noproxy".into(), "*".into()],
        ProxyMode::Custom { url, bypass } => {
            let noproxy = bypass
                .split(',')
                .map(str::trim)
                .filter(|part| !part.is_empty() && *part != "<local>")
                .collect::<Vec<_>>()
                .join(",");
            let mut args = vec!["--proxy".into(), url];
            if !noproxy.is_empty() {
                args.push("--noproxy".into());
                args.push(noproxy);
            }
            args
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    #[test]
    fn parse_accepts_http_and_socks5() {
        assert!(parse_proxy_url("http://127.0.0.1:7890").is_ok());
        assert!(parse_proxy_url("socks5://127.0.0.1:1080").is_ok());
        assert!(parse_proxy_url("ftp://127.0.0.1:1080").is_err());
        assert!(parse_proxy_url("http://").is_err());
        assert!(parse_proxy_url("socks5://127.0.0.1:1080 extra").is_err());
    }

    #[test]
    fn custom_mode_requires_url() {
        assert!(validate_network_proxy(&json!({
            "networkProxy": { "mode": "custom" }
        }))
        .is_err());
        assert!(validate_network_proxy(&json!({
            "networkProxy": { "mode": "custom", "url": "socks5://127.0.0.1:1080" }
        }))
        .is_ok());
        assert!(validate_network_proxy(&json!({
            "networkProxy": { "mode": "direct" }
        }))
        .is_ok());
    }

    #[test]
    fn curl_args_follow_applied_settings() {
        apply_from_settings(Some(&json!({
            "networkProxy": { "mode": "custom", "url": "socks5://127.0.0.1:1080" }
        })));
        let args = curl_proxy_args();
        assert_eq!(args[0], "--proxy");
        assert_eq!(args[1], "socks5://127.0.0.1:1080");
        assert!(args.contains(&"--noproxy".to_string()));

        apply_from_settings(Some(&json!({
            "networkProxy": { "mode": "direct" }
        })));
        assert_eq!(curl_proxy_args(), vec!["--noproxy", "*"]);

        apply_from_settings(Some(&json!({})));
        assert!(curl_proxy_args().is_empty());
    }
}
